ci/travis/build-docker-images.py

```diff
@@ -0,0 +1,260 @@
+            except Exception as e:
+                print(f"FAILURE with error {e}")
+
+            if len(DOCKER_CLIENT.api.images(tagged_name)) == 0:
+                print(f"ERROR building: {tagged_name} & error below:")
+                print(full_output)
+                if (i == 1):
+                    raise Exception("FAILED TO BUILD IMAGE")
+                print("TRYING AGAIN")
+            else:
+                break
+
+        print("BUILT: ", tagged_name)
+        built_images.append(tagged_name)
+    return built_images
+
+
+def copy_wheels():
+    root_dir = _get_root_dir()
+    wheel = _get_wheel_name()
+    source = os.path.join(root_dir, ".whl", wheel)
+    ray_dst = os.path.join(root_dir, "docker/ray/.whl/")
+    ray_dep_dst = os.path.join(root_dir, "docker/ray-deps/.whl/")
+    os.makedirs(ray_dst, exist_ok=True)
+    shutil.copy(source, ray_dst)
+    os.makedirs(ray_dep_dst, exist_ok=True)
+    shutil.copy(source, ray_dep_dst)
+
+
+def build_or_pull_base_images(is_docker_affected: bool) -> List[str]:
+    """Returns images to tag and build"""
+    _ = DOCKER_CLIENT.api.pull(
+        repository="rayproject/base-deps", tag="nightly")
+
+    age = DOCKER_CLIENT.api.inspect_image("rayproject/base-deps:nightly")[
+        "Created"]
+    short_date = datetime.datetime.strptime(age.split("T")[0], "%Y-%m-%d")
+    is_stale = (
+        datetime.datetime.now() - short_date) > datetime.timedelta(days=14)
+
+    if is_stale or is_docker_affected or _release_build():
+        for image in ["base-deps", "ray-deps"]:
+            _build_cpu_gpu_images(image)
+        return True
+    else:
+        print("Just pulling images!")
+        _ = DOCKER_CLIENT.api.pull(
+            repository="rayproject/base-deps", tag="nightly-cpu")
+        _ = DOCKER_CLIENT.api.pull(
+            repository="rayproject/base-deps", tag="nightly-gpu")
+
+        _ = DOCKER_CLIENT.api.pull(
+            repository="rayproject/ray-deps", tag="nightly-gpu")
+        _ = DOCKER_CLIENT.api.pull(
+            repository="rayproject/ray-deps", tag="nightly-cpu")
+        return False
+
+
+def build_ray():
+    return _build_cpu_gpu_images("ray")
+
+
+def build_ray_ml():
+    root_dir = _get_root_dir()
+    requirement_files = glob.glob(
+        f"{_get_root_dir()}/python/requirements*.txt")
+    for fl in requirement_files:
+        shutil.copy(fl, os.path.join(root_dir, "docker/ray-ml/"))
+    ray_ml_images = _build_cpu_gpu_images("ray-ml")
+    for img in ray_ml_images:
+        tag = img.split(":")[-1]
+        DOCKER_CLIENT.api.tag(
+            image=img, repository="rayproject/autoscaler", tag=tag)
+
+
+# For non-release builds, push "nightly" & "sha"
+# For release builds, push "nightly" & "latest" & "x.x.x"
+def push_and_tag_images(push_base_images: bool):
+    if _merge_build():
+        docker_password = os.environ.get("DOCKER_PASSWORD")
+        assert docker_password is not None, "DOCKER_PASSWORD not set."
```

Comment:
do we want to use `assert not docker_password` in the (unlikely) case that DOCKER_PASSWORD is set to an empty string?

Reply:
That would be a good fix!

Reply:
https://github.com/ray-project/ray/pull/11352/files#diff-b1c4770335cda67b7234504d0c7f899d41cd0042f172edb6f4f6d135325dee48R184